apps/src/lib/node/ledger/events/log.rs

```diff
@@ -0,0 +1,893 @@
+                    ControlFlow::Break(()) => false,
+                }
+            })
+            // build vec of new log nodes, all pointing to a null next node
+            .map(|n| {
+                Arc::new(LogNode {
+                    entry: n.entry.clone(),
+                    next: None,
+                })
+            })
+            .collect::<Vec<_>>()
+            // iterate the vec in reverse order, to link the nodes together in
+            // the correct order, e.g.: next <- head
+            .into_iter()
+            .rev()
+            .reduce(|next, mut head| {
+                Arc::get_mut(&mut head)
+                    .expect("There is only one live instance of this Arc")
+                    .next = Some(next);
+                head
+            })
+            .map(|head| EventLogSnapshot {
+                head,
+                num_events: total_events,
+                oldest_height,
+            })
+    }
+
+    /// Add a new entry to the log.
+    fn add(&self, entry: LogEntry) {
+        // do not log zero length event vecs
+        if entry.events.is_empty() {
+            return;
+        }
+
+        // update the log head
+        let (head, events, diff) = {
+            let mut log = self.inner.lock.write().unwrap();
+            let height_respects_log_invariant = {
+                log.head
+                    .as_ref()
+                    .map(|head| entry.block_height > head.entry.block_height)
+                    // if the log is empty, we can add whatever height
+                    // we want to the log
+                    .unwrap_or_else(|| {
+                        log.oldest_height = entry.block_height;
+                        true
+                    })
+            };
+            if !height_respects_log_invariant {
+                // drop entries whose block height is not
+                // higher than the log head's height
+                return;
+            }
+            let height_diff = entry.block_height.0 - log.oldest_height.0;
+            log.num_events += entry.events.len();
+            log.head = Some(Arc::new(LogNode {
+                entry,
+                next: log.head.take(),
+            }));
+            let new_head = log.head.clone();
+            (new_head, log.num_events, height_diff)
+        };
+
+        // notify all event listeners
+        self.inner.notifier.notify(usize::MAX);
+
+        // we don't need to hold a lock to check
+        // if the log needs to be pruned
+        self.prune(head, events, diff);
+    }
+}
+
+impl EventLogInnerMux {
+    /// Modifies the state of the [`EventLogInnerMux`] with the provided
+    /// [`EventLogSnapshot`].
+    fn install_snapshot(&mut self, snapshot: Option<EventLogSnapshot>) {
+        if let Some(snapshot) = snapshot {
+            self.oldest_height = snapshot.oldest_height;
+            self.num_events = snapshot.num_events;
+            self.head = Some(snapshot.head);
+        } else {
```

Comment:
If you derive `Default` on this type, the `else` can shorten to `*self = Default::default;`